File: programmers/힙/이중우선순위큐.java
import java.util.*;

class Solution {
    public int[] solution(String[] operations) {
        // <값, 빈도수> 저장
        TreeMap<Integer, Integer> map = new TreeMap<>();
        
        for (String op : operations) {
            String[] parts = op.split(" ");
            String cmd = parts[0];
            int num = Integer.parseInt(parts[1]);
            
            if (cmd.equals("I")) {
                // 삽입
                map.put(num, map.getOrDefault(num, 0) + 1);
                
            } else if (!map.isEmpty()) {
                // 삭제 연산 (빈 큐면 무시)
                int key = (num == 1)
                    ? map.lastKey()   // 최댓값
                    : map.firstKey(); // 최솟값
                
                // 빈도수 감소 또는 키 제거
                int cnt = map.get(key);
                if (cnt == 1) {
                    map.remove(key);
                } else {
                    map.put(key, cnt - 1);
                }
            }
        }
        
        // 결과 반환
        if (map.isEmpty()) {
            return new int[]{0, 0};
        } else {
            return new int[]{map.lastKey(), map.firstKey()};
        }
    }
}

File: programmers/힙/이중우선순위큐.py
import heapq

def solution(operations):    
    # 각 숫자에 고유한 ID 값을 부여하여 삭제 여부 추적
    id_idx = 0
    deleted_id = []
    max_heap = []
    min_heap = []
    for op in operations:
        op_str, num = op.split(" ")
        num = int(num)
        if op_str == "I":
            heapq.heappush(max_heap, (-num, id_idx))
            heapq.heappush(min_heap, (num, id_idx))
            deleted_id.append(False)
            id_idx += 1
        else:
            if num == 1:
                # 삭제된 ID라면 계속 꺼냄
                while max_heap and deleted_id[max_heap[0][1]]:
                    heapq.heappop(max_heap)
                if max_heap:
                    val, idx = heapq.heappop(max_heap)
                    deleted_id[idx] = True
            elif num == -1:
                # 삭제된 ID라면 계속 꺼냄
                while min_heap and deleted_id[min_heap[0][1]]: 
                    heapq.heappop(min_heap)
                if min_heap:
                    val, idx = heapq.heappop(min_heap)
                    deleted_id[idx] = True
    
    while max_heap and deleted_id[max_heap[0][1]]:
        heapq.heappop(max_heap)
    while max_heap and deleted_id[min_heap[0][1]]:
        heapq.heappop(min_heap)
    
    if not max_heap or not min_heap:
        return [0, 0]
    return [-max_heap[0][0], min_heap[0][0]]